pkg/runit/servicebuilder.go

```diff
@@ -18,51 +18,85 @@
 package runit
 
 import (
-	"bytes"
 	"fmt"
-	"io"
+	"io/ioutil"
 	"os"
-	"os/exec"
-	"path"
+	"path/filepath"
 
 	"github.com/square/p2/pkg/util"
 
 	"gopkg.in/yaml.v2"
 )
 
-type SBTemplate struct {
-	name       string
-	runClauses map[string]map[string][]string
+type ServiceTemplate struct {
+	Run      []string `yaml:"run"`
+	Log      []string `yaml:"log,omitempty"`
+	Sleep    int      `yaml:"sleep,omitempty"`
+	LogSleep int      `yaml:"logsleep,omitempty"`
 }
 
-type SBTemplateEntry struct {
-	baseCmd []string
-}
+func (s ServiceTemplate) RunScript() ([]byte, error) {
+	if len(s.Run) == 0 {
+		return nil, util.Errorf("empty run script")
+	}
+
+	args, err := yaml.Marshal(s.Run)
+	if err != nil {
+		return nil, err
+	}
 
-func NewSBTemplate(name string) *SBTemplate {
-	return &SBTemplate{
-		name:       name,
-		runClauses: make(map[string]map[string][]string),
+	// default sleep to reduce spinning on a broken run script
+	sleep := 2
+	if s.Sleep != 0 {
+		sleep = s.Sleep
 	}
-}
 
-func (template *SBTemplate) AddEntry(app string, run []string) {
-	template.runClauses[app] = map[string][]string{"run": run}
+	ret := fmt.Sprintf(`#!/usr/bin/ruby
+$stderr.reopen(STDOUT)
+require 'yaml'
+sleep %v
+exec *YAML.load(DATA.read)
+__END__
+%s`, sleep, args)
+	return []byte(ret), nil
 }
 
-func (template *SBTemplate) Write(out io.Writer) error {
-	b, err := yaml.Marshal(template.runClauses)
+func (s ServiceTemplate) LogScript() ([]byte, error) {
+	if len(s.Log) == 0 {
+		// use a default log script that makes a logdir, chowns it and execs
+		// svlogd into it
+		return []byte(`#!/usr/bin/ruby
+require 'fileutils'
+FileUtils.mkdir_p('./main')
+FileUtils.chown('nobody', 'nobody', './main')
+sleep 2
+exec('chpst', '-unobody', 'svlogd', '-tt', './main')
+`), nil
+	}
+
+	args, err := yaml.Marshal(s.Log)
 	if err != nil {
-		return util.Errorf("Could not marshal servicebuilder template %s as YAML: %s", template.name, err)
+		return nil, err
 	}
-	_, err = out.Write(b)
-	return err
+
+	sleep := 2
+	if s.LogSleep != 0 {
+		sleep = s.LogSleep
+	}
+
+	ret := fmt.Sprintf(`#!/usr/bin/ruby
+require 'yaml'
+sleep %v
+exec *YAML.load(DATA.read)
+__END__
+%s`, sleep, args)
```

Comment:
not quite, the run script reopens stderr where as the log script doesn't, which is why I've never copied them over. I'm not sure if reopening stderr would have any impact on svlogd, i'll check